src/base/settingsstorage.cpp

```diff
@@ -236,7 +194,94 @@ void SettingsStorage::removeValue(const QString &key)
     QWriteLocker locker(&m_lock);
     if (m_data.contains(realKey)) {
         m_dirty = true;
-        m_timer.start();
         m_data.remove(realKey);
+        m_timer.start();
+    }
+}
+
+void TransactionalSettings::read(QVariantHash &data)
+{
+#ifdef QSETTINGS_SYNC_IS_SAVE
+    serialize(m_name, data);
+#else
+    bool writeBackNeeded = false;
+    QString newPath = serialize(m_name + QLatin1String("_new"), data);
+    if (!newPath.isEmpty()) {// "_new" file is NOT empty
+        // This means that the PC closed either due to power outage
+        // or because the disk was full. In any case the settings weren't transfered
+        // in their final position. So assume that qbittorrent_new.ini/qbittorrent_new.conf
+        // contains the most recent settings.
+        Logger::instance()->addMessage(QObject::tr("Detected unclean program exit. Using fallback file to restore settings."), Log::WARNING);
+        writeBackNeeded = true;
+    }
+    else {
+        serialize(m_name, data);
+    }
+
+    Utils::Fs::forceRemove(newPath);
+
+    if (writeBackNeeded)
+        write(data);
+#endif
+}
+
+bool TransactionalSettings::write(const QVariantHash &data)
+{
+#ifdef QSETTINGS_SYNC_IS_SAVE
+    serialize(m_name, data);
+#else
+    // QSettings delete the file before writing it out. This can result in problems
+    // if the disk is full or a power outage occurs. Those events might occur
+    // between deleting the file and recreating it. This is a safety measure.
+    // Write everything to qBittorrent_new.ini/qBittorrent_new.conf and if it succeeds
+    // replace qBittorrent.ini/qBittorrent.conf with it.
+    QString newPath = serialize(m_name + QLatin1String("_new"), data);
+    if (newPath.isEmpty()) {
+        Utils::Fs::forceRemove(newPath);
+        return false;
+    }
+
+    QString finalPath = newPath;
+    int index = finalPath.lastIndexOf("_new", -1, Qt::CaseInsensitive);
+    finalPath.remove(index, 4);
+    Utils::Fs::forceRemove(finalPath);
+    QFile::rename(newPath, finalPath);
+#endif
+    return true;
+}
+
+QString TransactionalSettings::serialize(const QString &name, QVariantHash &data)
+{
+    SettingsPtr settings {createSettings(name)};
+
+    if (settings->allKeys().isEmpty())
+        return QString();
+
+    QStringList keys = settings->allKeys();
+
+    // Copy everything into memory. This means even keys inserted in the file manually
+    // or that we don't touch directly in this code(eg disabled by ifdef). This ensures
+    // that they will be copied over when save our settings to disk.
+    foreach (const QString &key, keys)
+        data.insert(key, settings->value(key));
+
+    return settings->fileName();
+}
+
+QString TransactionalSettings::serialize(const QString &name, const QVariantHash &data)
```

Comment:
You're doing that! You keep using overloaded functions incorrectly.
You're in a strange state. You know many advanced things, but you don't know some of the basics. I don't want to say that you're right from a technical point of view (I see it that way). But remember the story. Why do you have function overloading? To do the same thing with different things. But your methods are doing different things (one stores data, the other reads them). I will never agree with this approach.